File: backend/app/schemas/analytics.py
from __future__ import annotations

import enum
import uuid
from datetime import date

from pydantic import BaseModel


class StatusTag(enum.StrEnum):
    """Planned-vs-actual comparison status for a task."""

    DONE = "done"
    PARTIAL = "partial"
    SKIPPED = "skipped"
    UNPLANNED = "unplanned"


class TaskComparison(BaseModel):
    """Single task planned-vs-actual comparison."""

    task_id: uuid.UUID
    task_title: str
    planned_hours: float
    actual_hours: float
    status: StatusTag


class PlannedVsActualSummary(BaseModel):
    """Aggregate totals for the day."""

    total_planned_hours: float
    total_actual_hours: float
    done_count: int
    partial_count: int
    skipped_count: int
    unplanned_count: int


class PlannedVsActualResponse(BaseModel):
    """Full response for the planned-vs-actual endpoint."""

    date: date
    tasks: list[TaskComparison]
    summary: PlannedVsActualSummary

File: frontend/src/types/analytics.ts
export type StatusTag = 'done' | 'partial' | 'skipped' | 'unplanned'

export interface TaskComparison {
  task_id: string
  task_title: string
  planned_hours: number
  actual_hours: number
  status: StatusTag
}

export interface PlannedVsActualSummary {
  total_planned_hours: number
  total_actual_hours: number
  done_count: number
  partial_count: number
  skipped_count: number
  unplanned_count: number
}

export interface PlannedVsActualResponse {
  date: string
  tasks: TaskComparison[]
  summary: PlannedVsActualSummary
}

export interface ProjectWeeklyStats {
  project_id: string
  project_name: string
  project_color: string
  planned_hours: number
  actual_hours: number
  accuracy_pct: number
}

export interface WeeklyStatsSummary {
  total_planned_hours: number
  total_actual_hours: number
  average_accuracy_pct: number
}

export interface WeeklyStatsResponse {
  week_start: string
  week_end: string
  projects: ProjectWeeklyStats[]
  summary: WeeklyStatsSummary
}

export interface WeeklyChartEntry {
  name: string
  planned: number
  actual: number
  color: string
}
